File: backend/app/analysis/chips.py
"""筹码分布(成本分布 · 估算)—— 与 frontend/lib/chips.ts 保持一致。

用历史日 K 的量价堆积近似筹码:成交量在 [low, high] 均匀摊到价格桶 + 时间衰减。
获利比例 = 现价之下筹码占比;支撑/压力 = 现价上下最大筹码峰。估算参考,非真实持仓成本。
"""
from __future__ import annotations

from ..models import Bar


def compute_chips(bars: list[Bar], price: float, n_bins: int = 60, half_life: int = 60) -> dict | None:
    data = [b for b in bars if b.high > 0 and b.low > 0 and b.volume > 0]
    if len(data) < 10 or not price or price <= 0:
        return None

    lo = min(b.low for b in data)
    hi = max(b.high for b in data)
    if hi <= lo:
        return None
    step = (hi - lo) / n_bins
    bins = [0.0] * n_bins

    decay = 0.5 ** (1.0 / max(1, half_life))
    n = len(data)
    for i, b in enumerate(data):
        w = decay ** (n - 1 - i)
        vol = b.volume * w
        a = max(0, int((b.low - lo) / step))
        z = min(n_bins - 1, int((b.high - lo) / step))
        per = vol / (z - a + 1)
        for k in range(a, z + 1):
            bins[k] += per

    total = sum(bins) or 1.0
    levels = [{"price": lo + (i + 0.5) * step, "volume": v, "profit": (lo + (i + 0.5) * step) <= price}
              for i, v in enumerate(bins)]

    profit_ratio = sum(l["volume"] for l in levels if l["profit"]) / total
    avg_cost = sum(l["price"] * l["volume"] for l in levels) / total

    support = resistance = None
    s_v = r_v = 0.0
    for l in levels:
        if l["price"] < price and l["volume"] > s_v:
            s_v, support = l["volume"], l["price"]
        if l["price"] > price and l["volume"] > r_v:
            r_v, resistance = l["volume"], l["price"]

    low_cut, high_cut = total * 0.05, total * 0.95
    acc = 0.0
    c_low, c_high = levels[0]["price"], levels[-1]["price"]
    for l in levels:
        before = acc
        acc += l["volume"]
        if before < low_cut <= acc:
            c_low = l["price"]
        if before < high_cut <= acc:
            c_high = l["price"]
            break

    if profit_ratio >= 0.7:
        note = "高获利盘,上方抛压偏重。"
    elif profit_ratio <= 0.2:
        note = "深度套牢盘,反弹易遇压。"
    else:
        note = "获利与套牢盘相对均衡。"

    return {
        "price": price, "profit_ratio": profit_ratio, "avg_cost": avg_cost,
        "support": support, "resistance": resistance,
        "conc90": {"low": c_low, "high": c_high}, "levels": levels, "note": note,
    }

File: frontend/lib/chips.ts
// 筹码分布(成本分布)—— 对标富途「筹码分布 / 获利比例 / 支撑位·压力位」。
//
// 方法论:券商筹码分布基于真实持仓换手「移动」筹码;免费行情无持仓数据,
// 这里用历史日 K 估算 *成交量在价格上的堆积*:
//   · 每根 K 线把成交量在 [low, high] 区间内均匀摊到价格桶
//   · 时间衰减(越近权重越高),近似换手对筹码的稀释
//   · 获利比例 = 当前价之下(成本低于现价)的筹码占比
//   · 支撑位 = 现价之下最大的筹码峰;压力位 = 现价之上最大的筹码峰
// 为估算参考,非真实持仓成本。

import type { Bar } from "./datasource";

export interface ChipLevel { price: number; volume: number; profit: boolean }
export interface ChipDist {
  levels: ChipLevel[];          // 价格从低到高的筹码柱
  price: number;                // 现价
  profitRatio: number;          // 获利比例 0..1
  avgCost: number;              // 平均成本
  support: number | null;       // 支撑位(下方筹码峰)
  resistance: number | null;    // 压力位(上方筹码峰)
  conc90: { low: number; high: number }; // 90% 成本集中区间
  note: string;
}

export function computeChips(bars: Bar[], price: number, nBins = 60, halfLife = 60): ChipDist | null {
  const data = bars.filter((b) => b.high > 0 && b.low > 0 && b.volume > 0);
  if (data.length < 10 || !(price > 0)) return null;

  const lo = Math.min(...data.map((b) => b.low));
  const hi = Math.max(...data.map((b) => b.high));
  if (!(hi > lo)) return null;
  const step = (hi - lo) / nBins;
  const bins = new Array(nBins).fill(0);

  // 时间衰减:最近一根权重 1,越往前按半衰期衰减
  const decay = Math.pow(0.5, 1 / Math.max(1, halfLife));
  const n = data.length;
  for (let i = 0; i < n; i++) {
    const b = data[i];
    const w = Math.pow(decay, n - 1 - i);
    const vol = b.volume * w;
    const a = Math.max(0, Math.floor((b.low - lo) / step));
    const z = Math.min(nBins - 1, Math.floor((b.high - lo) / step));
    const span = z - a + 1;
    const per = vol / span;
    for (let k = a; k <= z; k++) bins[k] += per;
  }

  const total = bins.reduce((s, v) => s + v, 0) || 1;
  const levels: ChipLevel[] = bins.map((v, i) => {
    const p = lo + (i + 0.5) * step;
    return { price: p, volume: v, profit: p <= price };
  });

  const profitRatio = levels.filter((l) => l.profit).reduce((s, l) => s + l.volume, 0) / total;
  const avgCost = levels.reduce((s, l) => s + l.price * l.volume, 0) / total;

  // 支撑/压力:现价上下各自的最大筹码峰
  let support: number | null = null, resistance: number | null = null, sV = 0, rV = 0;
  for (const l of levels) {
    if (l.price < price && l.volume > sV) { sV = l.volume; support = l.price; }
    if (l.price > price && l.volume > rV) { rV = l.volume; resistance = l.price; }
  }

  // 90% 成本集中区间(去掉两端各 5% 筹码)
  const sortedByPrice = [...levels].sort((a, b) => a.price - b.price);
  const lowCut = total * 0.05, highCut = total * 0.95;
  let acc = 0, cLow = sortedByPrice[0].price, cHigh = sortedByPrice[sortedByPrice.length - 1].price;
  for (const l of sortedByPrice) {
    const before = acc; acc += l.volume;
    if (before < lowCut && acc >= lowCut) cLow = l.price;
    if (before < highCut && acc >= highCut) { cHigh = l.price; break; }
  }

  const note =
    profitRatio >= 0.7 ? "高获利盘,上方抛压偏重,留意获利了结。"
      : profitRatio <= 0.2 ? "深度套牢盘,上方套牢需化解,反弹易遇压。"
        : "获利与套牢盘相对均衡。";

  return { levels, price, profitRatio, avgCost, support, resistance, conc90: { low: cLow, high: cHigh }, note };
}
